feat: caching of generated images

pzerelles opened the pull request:
Closes https://github.com/JonasKruckenberg/imagetools/issues/522

- **Quick Checklist**

* [x] I have read [the contributing guidelines](../CONTRIBUTING.md)
* [x] I have written new tests, as applicable (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] I have added a changeset, if applicable

- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature

- **What is the new behavior (if this is a feature change)?**
Generated images can be cached to skip transformation of unchanged images.

- **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this
  PR?)
No

- **Other information**:

srueg commented on packages/vite/src/index.ts:
```diff
@@ -171,17 +259,46 @@ export function imagetools(userOptions: Partial<VitePluginOptions> = {}): Plugin
+              }
+            } catch (e) {
+              console.debug(`deleting invalid cache dir ${dirent.name}`)
+              await rm(`${cacheDir}/${dirent.name}`, { recursive: true })
```
```suggestion
              await rm(cacheDir, { recursive: true })
```

AdamMerrifield commented:
Are there any other issues with this other than the conflicts that need to be resolved? I have tested it locally and it is working well for me.

benmccann commented:
I was hoping to get caching support in Vite first, but I don't think that's going to happen soon enough, so I plan to get this PR in. I won't be able to look at it for the next week, but hopefully after that.

benmccann commented:
@pzerelles heads up that this PR needs a rebase

benmccann commented on .changeset/afraid-books-notice.md:
```diff
@@ -0,0 +1,5 @@
+'vite-imagetools': patch
+---
+
+fix: use URI-decoded pathname when emitting file
```
can we split the URI decoding stuff out into a separate PR?

pzerelles replied:
Yes, but it will look different before and after caching. Do you want the URI decoding before the cache is integrated or after?

pzerelles replied:
I created PR #689 that can be applied before the caching is merged.
If you decide to merge this PR first, I will rebase the cache PR and add the necessary changes.
If you decide to merge the cache PR first, I will rebase this PR and add the necessary changes.

Aragur commented:
Awesome PR, I was looking for a way to cache my pictures. Pipeline takes 30 minutes because I have many images.
Are there any updates on this PR, when will it be merged? Is there anything I can do to help?

benmccann commented:
This one has a merge conflict I think from merging your other PR. Can you update it? Thanks!

benmccann commented on docs/guide/caching.md:
```diff
@@ -0,0 +1,39 @@
+
+## How to enable caching
+
+To enable caching, the cache directory has to be configured.
```
maybe we should rename it from `cacheDir` to `cache` at the same time as `cacheDir: false` sounds a bit funny. or we could have:
```
cache: {
  enabled: true,
  dir: './node_modules/.cache/imagetools'
}
```

pzerelles replied:
Good idea. Can maybe work that in later.

julien-blanchon commented:
Is this stable enouth to be use in dev ? I have almost 20min of build time on Vercel (almost 40 image to be optimize by imagetools) ... this is a nightmare for development. I'm really thinking to stop using imagetools because of that ?

pzerelles commented:
> Is this stable enouth to be use in dev ? I have almost 20min of build time on Vercel (almost 40 image to be optimize by imagetools) ... this is a nightmare for development. I'm really thinking to stop using imagetools because of that ?

I've been using it in dev and production for half a year already.

julien-blanchon commented:
> > Is this stable enouth to be use in dev ? I have almost 20min of build time on Vercel (almost 40 image to be optimize by imagetools) ... this is a nightmare for development. I'm really thinking to stop using imagetools because of that ?
> 
> I've been using it in dev and production for half a year already.

How did you specify pnpm/npm/yarn to fetch this branch ?
Any suggestion for Vercel, I suppose it work out the box

pzerelles commented:
> > > Is this stable enouth to be use in dev ? I have almost 20min of build time on Vercel (almost 40 image to be optimize by imagetools) ... this is a nightmare for development. I'm really thinking to stop using imagetools because of that ?
> 
> > 
> 
> > I've been using it in dev and production for half a year already.
> 
> 
> 
> How did you specify pnpm/npm/yarn to fetch this branch ?
> 
> Any suggestion for Vercel, I suppose it work out the box

I use a personal build created from my forked repository. But I hope this will be merged really soon. We are working on it.

benmccann commented on packages/vite/src/utils.ts:
```diff
@@ -29,3 +28,18 @@ function hash(keyParts: Array<string | NodeJS.ArrayBufferView>) {
   return hash.digest('hex')
 }
+
+export const generateCacheID = (path: string) => createHash('sha1').update(path).digest('hex')
```
This looks almost exactly the same as the hash function right above it. Could we figure out how to just have one function for this?

pzerelles replied:
I removed "generateCacheID" in favor of using your "hash" function.

benmccann commented on packages/vite/src/utils.ts:
```diff
@@ -29,3 +28,18 @@ function hash(keyParts: Array<string | NodeJS.ArrayBufferView>) {
+
+export const generateCacheID = (path: string) => createHash('sha1').update(path).digest('hex')
+
+export const checksumFile = (algorithm: string, path: string) => {
```
Do we need a new `checksumFile` method or can we reuse `generateImageID`, which does something similar already? Also, this is more of a hash than a checksum, so perhaps we could rename it

pzerelles replied:
I renamed the function and the property in the cache metadata file.

benmccann commented on packages/vite/src/index.ts:
```diff
@@ -172,17 +263,46 @@ export function imagetools(userOptions: Partial<VitePluginOptions> = {}): Plugin
+    },
+
+    async buildEnd(error) {
+      if (!error && cacheOptions.enabled && cacheOptions.retention && viteConfig.command !== 'serve') {
```
```suggestion
      // clear expired cache entries
      if (!error && cacheOptions.enabled && cacheOptions.retention && viteConfig.command !== 'serve') {
```

pzerelles replied:
Added

benmccann commented on docs/guide/caching.md:
```diff
@@ -0,0 +1,46 @@
+
+When an image is no longer there or the transformation parameters change, the previously
+cached images will be removed after a configurable retention period.
+The default retention is 86400 seconds. A value of 0 will disable this mechanism.
```
this seems kind of short. perhaps it should be disabled by default?
```suggestion
The default retention is 1 day (86400 seconds). A value of 0 will disable this mechanism.
```

pzerelles replied:
Changed

pzerelles replied:
Could you please have a look at `docs/interfaces/vite_src_types.VitePluginOptions.md`?

benmccann commented on packages/vite/src/__tests__/main.test.ts:
```diff
@@ -7,7 +7,15 @@ import { OutputAsset, OutputChunk, RollupOutput } from 'rollup'
+import { createBasePath, generateCacheID } from '../utils'
+import { readFile, rm, utimes } from 'fs/promises'
+
+const processPath = process.cwd()
```
I think `cwd` would be a clearer name. I know we used that variable name in the sveltekit codebase in the path so it would help keep them aligned
```suggestion
const cwd = process.cwd()
```

pzerelles replied:
Changed

benmccann commented on packages/vite/src/index.ts:
```diff
@@ -51,7 +62,11 @@ export function imagetools(userOptions: Partial<VitePluginOptions> = {}): Plugin
   let basePath: string
 
-  const generatedImages = new Map<string, Sharp>()
+  const processPath = process.cwd()
```
```suggestion
  const cwd = process.cwd()
```

pzerelles replied:
Changed

benmccann commented:
It looks to me like if a user changes the way in which they want an image to be transformed that those changes won't get picked up because we prefer the cache.

I think we may be able to remove the need for an `index.json` file if we write a copy of the image file to the cache directory where the filename is the ID returned by generateImageID. Then we could almost use the logic that exists today, but instead of checking the in-memory `generatedImages` cache we could check the cache directory.

pzerelles commented:
> It looks to me like if a user changes the way in which they want an image to be transformed that those changes won't get picked up because we prefer the cache.
> 
> I think we may be able to remove the need for an `index.json` file if we write a copy of the image file to the cache directory where the filename is the ID returned by generateImageID. Then we could almost use the logic that exists today, but instead of checking the in-memory `generatedImages` cache we could check the cache directory.

The main purpose of the `index.json` file is to provide the metadata without reading the image. But this is only an advantage in dev mode, because the file will be read anyway for `emitFile` in build mode. I could try what you suggest.

pzerelles commented:
I refactored the whole PR to use the ID from `generateImageID`. Since user changes will also change that ID, the former problem is solved. I adjusted tests where necessary.

AdamMerrifield commented:
@pzerelles while using this, i have found that it is possible for 0 byte files to be stored in the cache directory if a build is cancelled while generating images. 

I sent a PR to resolve this issue https://github.com/pzerelles/imagetools/pull/1

benmccann commented on .changeset/quick-roses-fail.md:
```diff
@@ -0,0 +1,5 @@
+---
+'vite-imagetools': patch
```
```suggestion
'vite-imagetools': minor
```

benmccann commented on docs/guide/caching.md:
```diff
@@ -0,0 +1,46 @@
+
+## How to configure caching
+
+Caching is enabled by default and uses './node_modules/.cache/imagetools' as cache directory. 
```
```suggestion
Caching is enabled by default and uses './node_modules/.cache/imagetools' as the cache directory. 
```

benmccann commented on packages/vite/src/index.ts:
```diff
@@ -167,22 +199,44 @@ export function imagetools(userOptions: Partial<VitePluginOptions> = {}): Plugin
+          if (!metadata)
             throw new Error(`vite-imagetools cannot find image with id "${id}" this is likely an internal error`)
 
+          res.setHeader('Cache-Control', 'max-age=360000')
+
```
I know this was already present, but I'm not really sure why we had it. It seems like it just has the potential to cause issues if the underlying image changes. And with the other caching that this plugin has, serving the requests shouldn't be all that expensive. We could potentially add an etag header as an alternative
```suggestion
```

pzerelles replied:
I don't think it will cause issues, because every change to the original images or the transform options will change the image ID, too. But we can try without it.

benmccann commented on packages/vite/src/index.ts:
```diff
@@ -122,18 +133,39 @@ export function imagetools(userOptions: Partial<VitePluginOptions> = {}): Plugin
       for (const config of imageConfigs) {
-        const { transforms } = generateTransforms(config, transformFactories, srcURL.searchParams, logger)
-        const { image, metadata } = await applyTransforms(transforms, img.clone(), pluginOptions.removeMetadata)
+        const id = await generateImageID(srcURL, config, imageBuffer)
+        let image: Sharp | undefined
+        let metadata: ImageMetadata
+
+        if (cacheOptions.enabled && existsSync(`${cacheOptions.dir}/${id}`)) {
+          const imagePath = `${cacheOptions.dir}/${id}`
+          metadata = (await sharp(imagePath).metadata()) as ImageMetadata
+          metadata.imagePath = imagePath
+          const date = new Date()
+          utimes(imagePath, date, date)
+        } else {
+          const { transforms } = generateTransforms(config, transformFactories, srcURL.searchParams, logger)
+          const res = await applyTransforms(transforms, img, pluginOptions.removeMetadata)
+          metadata = res.metadata
+          if (cacheOptions.enabled) {
+            const imagePath = `${cacheOptions.dir}/${id}`
+            await writeFile(imagePath, await res.image.toBuffer())
+            metadata.imagePath = imagePath
+          } else {
+            image = res.image
+          }
+        }
```
I'm not sure how I feel about sticking `imagePath` on `metadata`. Could we do something like this instead?
```suggestion
        const id = await generateImageID(srcURL, config, imageBuffer)
        let image: Sharp | undefined
        let imagePath: string: undefined
        let metadata: ImageMetadata

        if (cacheOptions.enabled && existsSync(`${cacheOptions.dir}/${id}`)) {
          imagePath = `${cacheOptions.dir}/${id}`
          metadata = (await sharp(imagePath).metadata()) as ImageMetadata
          const date = new Date()
          utimes(imagePath, date, date)
        } else {
          const { transforms } = generateTransforms(config, transformFactories, srcURL.searchParams, logger)
          const res = await applyTransforms(transforms, img, pluginOptions.removeMetadata)
          metadata = res.metadata
          if (cacheOptions.enabled) {
            imagePath = `${cacheOptions.dir}/${id}`
            await writeFile(imagePath, await res.image.toBuffer())
          } else {
            image = res.image
          }
        }
```

benmccann commented on packages/vite/src/index.ts:
```diff
@@ -122,18 +133,39 @@ export function imagetools(userOptions: Partial<VitePluginOptions> = {}): Plugin
       for (const config of imageConfigs) {
-        const { transforms } = generateTransforms(config, transformFactories, srcURL.searchParams, logger)
-        const { image, metadata } = await applyTransforms(transforms, img.clone(), pluginOptions.removeMetadata)
+        const id = await generateImageID(srcURL, config, imageBuffer)
+        let image: Sharp | undefined
+        let metadata: ImageMetadata
+
+        if (cacheOptions.enabled && existsSync(`${cacheOptions.dir}/${id}`)) {
+          const imagePath = `${cacheOptions.dir}/${id}`
+          metadata = (await sharp(imagePath).metadata()) as ImageMetadata
+          metadata.imagePath = imagePath
```
I'm not sure how I feel about sticking `imagePath` on `metadata`. Could we do something like this instead?
```suggestion
        const id = await generateImageID(srcURL, config, imageBuffer)
        let image: Sharp | undefined
        let imagePath: string | undefined
        let metadata: ImageMetadata

        if (cacheOptions.enabled && existsSync(`${cacheOptions.dir}/${id}`)) {
          imagePath = `${cacheOptions.dir}/${id}`
          metadata = (await sharp(imagePath).metadata()) as ImageMetadata
```

benmccann commented on packages/vite/src/index.ts:
```diff
@@ -122,18 +133,39 @@ export function imagetools(userOptions: Partial<VitePluginOptions> = {}): Plugin
+          const res = await applyTransforms(transforms, img, pluginOptions.removeMetadata)
+          metadata = res.metadata
+          if (cacheOptions.enabled) {
+            const imagePath = `${cacheOptions.dir}/${id}`
+            await writeFile(imagePath, await res.image.toBuffer())
+            metadata.imagePath = imagePath
```
if we go with the suggestion above, this would need to change to:
```suggestion
            imagePath = `${cacheOptions.dir}/${id}`
            await writeFile(imagePath, await res.image.toBuffer())
```

benmccann commented on packages/vite/src/index.ts:
```diff
@@ -122,18 +133,39 @@ export function imagetools(userOptions: Partial<VitePluginOptions> = {}): Plugin
           const fileHandle = this.emitFile({
             name: basename(pathname, extname(pathname)) + `.${metadata.format}`,
-            source: await image.toBuffer(),
+            source: image ? await image.toBuffer() : await readFile(metadata.imagePath as string),
```
if we go with the suggestion above, this would need to change to:
```suggestion
            source: image ? await image.toBuffer() : await readFile(imagePath as string),
```

benmccann commented on docs/interfaces/vite_src_types.VitePluginOptions.md:
```diff
@@ -15,6 +15,8 @@
 - [include](vite_src_types.VitePluginOptions.md#include)
 - [removeMetadata](vite_src_types.VitePluginOptions.md#removemetadata)
 - [resolveConfigs](vite_src_types.VitePluginOptions.md#resolveconfigs)
+- [cacheRetention](vite_src_types.VitePluginOptions.md#cacheretention)
+- [cacheDir](vite_src_types.VitePluginOptions.md#cachedir)
```
these will need to be updated to match the new option names

benmccann replied:
these will need to be updated to match the new option names
```suggestion
- [cache](vite_src_types.VitePluginOptions.md#cache)
```

benmccann commented on packages/vite/src/__tests__/main.test.ts:
```diff
@@ -8,6 +8,8 @@ import { JSDOM } from 'jsdom'
 import sharp from 'sharp'
 import { afterEach, describe, test, expect, it, vi } from 'vitest'
 import { createBasePath } from '../utils'
+import { existsSync } from 'fs'
+import { rm, utimes, readdir } from 'fs/promises'
```
```suggestion
import { existsSync } from 'node:fs'
import { rm, utimes, readdir } from 'node:fs/promises'
```

benmccann commented on packages/vite/src/__tests__/main.test.ts:
```diff
@@ -458,6 +466,59 @@ describe('vite-imagetools', () => {
         expect(window.__IMAGE__).toHaveProperty('hasAlpha')
       })
     })
+    describe('cache.retention', () => {
```
```suggestion

    describe('cache.retention', () => {
```

benmccann commented on packages/vite/src/__tests__/main.test.ts:
```diff
@@ -458,6 +466,59 @@ describe('vite-imagetools', () => {
+        expect(existsSync(`${dir}/${image_300}`)).toBe(false)
+      })
+    })
+    describe('cache.dir', () => {
```
```suggestion

    describe('cache.dir', () => {
```

benmccann commented on docs/guide/caching.md:
```diff
@@ -0,0 +1,46 @@
+
+When an image is no longer there or the transformation parameters change, the previously
+cached images will be removed after a configurable retention period.
+The default retention is 1 day (86400 seconds). A value of 0 will disable this mechanism.
```
can we have this be `undefined` by default? 1 day seems too short

pzerelles replied:
I was thinking about also adding images to `generatedImages` during build and only remove the images if they are not there. If you have a CI/CD system, that does not restore the cache with the correct timestamps, you could set retention to 1 second and only unused images will be removed immediately.

pzerelles replied:
What effect should `undefined` have? Should that disable retention?

benmccann replied:
yeah

julien-blanchon commented:
Yayyy, will this be release soon ? How can I update in the meantime ?

benmccann commented:
Yes, it will be released before too long. It will be available when https://github.com/JonasKruckenberg/imagetools/pull/707 is merged. You can subscribe to that PR for updates

julien-blanchon commented:
Humm I don't know why but it's seems to not work with 7.0, and 
```
imagetools({
			cache: {
				enabled: true,
				dir: './node_modules/.cache/imagetools',
			}
		}),
```

In the `vite.config.ts`.

I'm using Sveltekit with a bunch of `?enhanced' imports

benmccann commented:
You'll need to wait for https://github.com/sveltejs/kit/pull/12055 for `@sveltejs/enhanced-img`. You shouldn't need to setup image tools.in your Vite config with `@sveltejs/enhanced-img`
